src/account/sync/input_selection.rs

```diff
@@ -41,36 +68,59 @@ pub fn select_input(target: u64, mut available_utxos: Vec<Input>) -> crate::Resu
     let mut selected_coins = Vec::new();
     let result = branch_and_bound(
         target,
-        &mut available_utxos,
+        &signature_locked_outputs,
         0,
         &mut selected_coins,
         0,
         &mut AtomicI64::new(MAX_INPUT_SELECTION_TRIES),
     );
 
-    let selected_balance = selected_coins.iter().fold(0, |acc, address| acc + address.balance);
+    let selected_balance = selected_coins.iter().fold(0, |acc, input| acc + input.output.amount);
     let remaining_value = if selected_balance >= target {
         selected_balance - target
     } else {
         0
     };
 
-    if result && selected_balance >= target && (remaining_value == 0 || remaining_value > DUST_ALLOWANCE_VALUE) {
+    if result
+        && selected_balance >= target
+        && selected_coins.len() <= max_inputs
+        && (remaining_value == 0 || remaining_value > DUST_ALLOWANCE_VALUE)
+    {
         Ok(selected_coins)
     } else {
         // If no match, Single Random Draw
-        Ok(single_random_draw(target, available_utxos))
+        let mut signature_locked_outputs_ = signature_locked_outputs.clone();
+        let mut dust_allowance_outputs_ = dust_allowance_outputs.clone();
+        signature_locked_outputs_.shuffle(&mut thread_rng());
+        dust_allowance_outputs_.shuffle(&mut thread_rng());
+        let mut inputs = single_draw(target, signature_locked_outputs_, dust_allowance_outputs_);
+        if inputs.len() > max_inputs {
```

Comment:
Don't we need to sort the dust_allowance_outputs?

Reply:
Yes